packages/server-wallet/src/protocols/challenge-submitter.ts

```diff
@@ -27,55 +34,56 @@ export class ChallengeSubmitter {
-    response: WalletResponse
-  ): Promise<void> {
+    response: WalletResponse,
+    tx: Transaction
+  ): Promise<WaitingFor | null> {
     const {targetChannelId: channelToLock} = objective.data;
 
-    await this.store.transaction(async tx => {
-      const channel = await this.store.getAndLockChannel(channelToLock, tx);
-      if (!channel) {
-        this.logger.warn(`No channel exists for channel ${channelToLock}`);
-        return;
-      }
+    const channel = await this.store.getAndLockChannel(channelToLock, tx);
+    if (!channel) {
+      throw new Error('Channel must exist');
+    }
 
-      const status = channel.adjudicatorStatus
-        ? channel.adjudicatorStatus.toResult().channelMode
-        : 'Open';
-      if (status !== 'Open') {
-        this.logger.warn('There is an existing challenge or the channel is finalized on chain');
-        return;
-      }
+    const status = channel.adjudicatorStatus
+      ? channel.adjudicatorStatus.toResult().channelMode
+      : 'Open';
+    if (status !== 'Open') {
+      this.logger.warn('There is an existing challenge or the channel is finalized on chain');
+      // TODO if the channel is finalized, the objective should probably terminate
+      // TODO if there is an existing challenge, we may still want to submit a new one
+      return WaitingFor.existingChallenge;
+    }
 
-      if (!channel.signingWallet) {
-        throw new Error(`No signing wallet fetched for channel ${channelToLock}`);
-      }
+    if (!channel.signingWallet) {
+      throw new Error(`No signing wallet fetched for channel ${channelToLock}`);
+    }
 
-      const existingRequest = await ChainServiceRequest.query(tx)
-        .where({channelId: channelToLock, request: 'challenge'})
-        .first();
+    const existingRequest = await ChainServiceRequest.query(tx)
+      .where({channelId: channelToLock, request: 'challenge'})
+      .first();
 
-      if (existingRequest) {
-        this.logger.warn('There is already an existing request', existingRequest);
-        return;
-      }
+    if (existingRequest) {
+      this.logger.warn('There is already an existing request', existingRequest);
+      return WaitingFor.existingChainServiceRequest;
+    }
 
-      await ChainServiceRequest.insertOrUpdate(channelToLock, 'challenge', tx);
+    await ChainServiceRequest.insertOrUpdate(channelToLock, 'challenge', tx);
 
-      if (channel.initialSupport.length === 0) {
-        this.logger.error(
-          {channelId: channel.channelId},
-          'There is no initial support stored for the channel'
-        );
-        await this.store.markObjectiveStatus(objective, 'failed', tx);
-        return;
-      }
+    if (channel.initialSupport.length === 0) {
+      this.logger.error(
+        {channelId: channel.channelId},
+        'There is no initial support stored for the channel'
+      );
+      await this.store.markObjectiveStatus(objective, 'failed', tx);
+      return null;
```

Comment:
Is there a reason we're using `null` not `undefined`?

Reply:
Yes: querying the database may return `null`. If we want to use `undefined`, I believe we would need to handle conversion from `null` to undefined. I felt that it was safer and easier to stick with `null`. Happy to discuss though.

Reply:
I think I'd prefer to keep `null` types restricted to the data layer only. We generally don't use the `null` type throughout our codebase and I could see a bug easily being introduced if someone adds a check like `o.waitingFor === undefined`.